cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/BrokerSetAwareGoal.java

```diff
@@ -0,0 +1,327 @@
+                                             recommendation);
+    }
+
+    // Whether the {@link com.linkedin.kafka.cruisecontrol.model.SortedReplicas} tracks replicas in descending order or not.
+    boolean tracksReplicasInReverseOrder = false;
+    // Whether the {@link com.linkedin.kafka.cruisecontrol.model.SortedReplicas} tracks only leader replicas or all replicas.
+    boolean tracksOnlyLeaderReplicas = false;
+
+    /*
+    Replicas using selection/priority/score functions can be filtered/ordered/scored to be picked for movement during balancing act.
+    If the Optimization Options requires movement of only immigrant replicas, then we pick immigrant replicas.
+    Similarly, if excluded topics are specified, then for optimizing efficiently, then we select only included topics' replicas.
+    The reason for picking the below selection functions is to abide by the Optimization Options.
+    Other functions like selecting leaders etc are not relevant for BrokerSetAwareGoal.
+    */
+    new SortedReplicasHelper().maybeAddSelectionFunc(ReplicaSortFunctionFactory.selectImmigrants(),
+                                                     optimizationOptions.onlyMoveImmigrantReplicas())
+                              .maybeAddSelectionFunc(ReplicaSortFunctionFactory.selectReplicasBasedOnExcludedTopics(excludedTopics),
+                                                     !excludedTopics.isEmpty())
+                              .trackSortedReplicasFor(replicaSortName(this, tracksReplicasInReverseOrder, tracksOnlyLeaderReplicas),
+                                                      clusterModel);
+
+    BrokerSetResolver brokerSetResolver = _balancingConstraint.brokerSetResolver();
+    _brokerSetResolutionHelper = new BrokerSetResolutionHelper(clusterModel, brokerSetResolver);
+    _brokersByBrokerSet = _brokerSetResolutionHelper.brokersByBrokerSetId();
+    _replicaToBrokerSetMappingPolicy = _balancingConstraint.replicaToBrokerSetMappingPolicy();
+  }
+
+  /**
+   * Update goal state.
+   * Sanity check: After completion of balancing / self-healing, confirm that replicas of each topic reside fully in a
+   * broker set.
+   *
+   * @param clusterModel        The state of the cluster.
+   * @param optimizationOptions Options to take into account during optimization.
+   */
+  @Override
+  protected void updateGoalState(ClusterModel clusterModel, OptimizationOptions optimizationOptions) throws OptimizationFailureException {
+    // Sanity check: No self-healing eligible replica should remain at a dead broker/disk.
+    GoalUtils.ensureNoOfflineReplicas(clusterModel, name());
+    // Sanity check: No replica should be moved to a broker, which used to host any replica of the same partition on its broken disk.
+    GoalUtils.ensureReplicasMoveOffBrokersWithBadDisks(clusterModel, name());
+    // Sanity check to confirm that the final distribution is broker set aware.
+    ensureBrokerSetAware(clusterModel, optimizationOptions);
```

Comment:
Just caught another issue during testing:

```
2022/05/21 10:18:49.621 WARN [OperationRunnable] [ServletSessionExecutor-2] [kafka-cruise-control] [] Received exception when trying to execute runnable for "Rebalance"
com.linkedin.kafka.cruisecontrol.exception.OptimizationFailureException: [BrokerSetAwareGoal] Topic ABTestMetricDashboardUserEvent is not brokerSet-aware.

```

The problem here is rebalance is done broker by broker. Let's say we have 10 brokers, and topics are randomly distributed across colors. After the rebalanceForBroker is called for broker 1, it's totally possible that there are still topics not resides within a single color, then the goal will fail with a high chance. I'm thinking maybe we should remove this check here.

Reply:
(Only if you agree with the above comments): also maybe a test for this case would be helpful?

Reply:
Did you change the configuration ``replica.to.broker.set.mapping.policy.class`` as ``com.linkedin.kafka.cruisecontrol.config.TopicNameHashBrokerSetMappingPolicy``

Reply:
Oh not yet. I was testing without your latest change this morning. So it's the previous version.

I'm wondering even I change the policy would it avoid this failure? The rebalance still happens broker by broker, and  brokerSetAware only satisfy after the rebalance for all brokers completed. Does that make sense or did I miss something?

Reply:
I tested it, it does work in above case. The ensureBrokerSetAware gets called after the goal is run as part of ``UpdateGoalState`` after it has finished rebalancing for all brokers.

Reply:
Oops I was wrong. Sorry about it! You're right that UpdateGoalState only happens after rebalance for each broker:

```
for (Broker broker : brokersToBalance(clusterModel)) {
  rebalanceForBroker(broker, clusterModel, optimizedGoals, optimizationOptions);
}
updateGoalState(clusterModel, optimizationOptions);
```

The failure might just because it moves replicas to it's original broker set, which in our case is still different colors.

That said, since by default Kafka does not guarantee topics created in same color, shall we make TopicNameHashBrokerSetMappingPolicy as the default policy? E.g. when other people/company use it without any extra infra level changes, it won't fail.

Reply:
Makes sense, the only thing is that hashing based policy does not guarantee evenness of distribution. Multiple topics can land on a single broker set as this is consistent hash based. However, the other goals will try to optimize the distribution so i guess that is fine. I will revise the PR with the default change and resubmit.

Reply:
Awesome. I'll have another test once you make that change. Thanks Mohit!